File: experiment-13/runs/language=go_model=opus-4.8-fast_prompt=TDD/rep1/mcp.go
// Package main — Brazilian Soccer MCP Server.
//
// mcp.go: The Model Context Protocol server core — a JSON-RPC 2.0 dispatcher
// over the loaded Dataset. It implements the MCP lifecycle (initialize, the
// initialized notification, ping) and tool surface (tools/list, tools/call),
// translating tool handler results into MCP content blocks. The transport
// (newline-delimited JSON over stdio) lives in main.go; Dispatch is kept
// transport-agnostic and pure so it can be unit-tested directly.
package main

import (
	"encoding/json"
)

// protocolVersion is the MCP revision this server speaks.
const protocolVersion = "2024-11-05"

// serverName/serverVersion identify this server in the initialize handshake.
const (
	serverName    = "brazilian-soccer-mcp"
	serverVersion = "1.0.0"
)

// JSON-RPC standard error codes.
const (
	codeParseError     = -32700
	codeInvalidRequest = -32600
	codeMethodNotFound = -32601
	codeInvalidParams  = -32602
	codeInternalError  = -32603
)

// Server holds the loaded dataset and the tool registry, dispatching JSON-RPC
// requests against them.
type Server struct {
	ds    *Dataset
	tools map[string]Tool
	order []string // tool names in catalog order, for tools/list
}

// NewServer builds a Server over the given dataset.
func NewServer(ds *Dataset) *Server {
	s := &Server{ds: ds, tools: map[string]Tool{}}
	for _, t := range Tools() {
		s.tools[t.Name] = t
		s.order = append(s.order, t.Name)
	}
	return s
}

// rpcRequest is an incoming JSON-RPC message. ID is kept as raw JSON so it can
// be echoed back exactly (number or string), and its absence marks a
// notification.
type rpcRequest struct {
	JSONRPC string          `json:"jsonrpc"`
	ID      json.RawMessage `json:"id"`
	Method  string          `json:"method"`
	Params  json.RawMessage `json:"params"`
}

type rpcError struct {
	Code    int    `json:"code"`
	Message string `json:"message"`
}

// Dispatch handles one raw JSON-RPC request and returns the marshaled response.
// The boolean is false when no response should be sent (notifications), or when
// the input could not be parsed at all as a message with an id to reply to.
func (s *Server) Dispatch(raw []byte) ([]byte, bool) {
	var req rpcRequest
	if err := json.Unmarshal(raw, &req); err != nil {
		return s.errorResponse(nil, codeParseError, "parse error"), true
	}
	isNotification := len(req.ID) == 0

	switch req.Method {
	case "initialize":
		return s.resultResponse(req.ID, s.initializeResult()), true
	case "notifications/initialized", "notifications/cancelled":
		return nil, false
	case "ping":
		return s.resultResponse(req.ID, map[string]any{}), true
	case "tools/list":
		return s.resultResponse(req.ID, s.toolsListResult()), true
	case "tools/call":
		return s.handleToolCall(req)
	default:
		if isNotification {
			return nil, false
		}
		return s.errorResponse(req.ID, codeMethodNotFound, "method not found: "+req.Method), true
	}
}

func (s *Server) initializeResult() map[string]any {
	return map[string]any{
		"protocolVersion": protocolVersion,
		"capabilities": map[string]any{
			"tools": map[string]any{},
		},
		"serverInfo": map[string]any{
			"name":    serverName,
			"version": serverVersion,
		},
		"instructions": "Query Brazilian soccer datasets: matches (Brasileirão, Copa do Brasil, Libertadores), team records, standings, head-to-head, and FIFA player data.",
	}
}

func (s *Server) toolsListResult() map[string]any {
	tools := make([]map[string]any, 0, len(s.order))
	for _, name := range s.order {
		t := s.tools[name]
		tools = append(tools, map[string]any{
			"name":        t.Name,
			"description": t.Description,
			"inputSchema": t.InputSchema,
		})
	}
	return map[string]any{"tools": tools}
}

// toolCallParams is the params shape for tools/call.
type toolCallParams struct {
	Name      string         `json:"name"`
	Arguments map[string]any `json:"arguments"`
}

func (s *Server) handleToolCall(req rpcRequest) ([]byte, bool) {
	var p toolCallParams
	if len(req.Params) > 0 {
		if err := json.Unmarshal(req.Params, &p); err != nil {
			return s.errorResponse(req.ID, codeInvalidParams, "invalid params: "+err.Error()), true
		}
	}
	tool, ok := s.tools[p.Name]
	if !ok {
		return s.toolError(req.ID, "unknown tool: "+p.Name), true
	}
	if p.Arguments == nil {
		p.Arguments = map[string]any{}
	}
	text, err := tool.Handler(s.ds, p.Arguments)
	if err != nil {
		return s.toolError(req.ID, err.Error()), true
	}
	return s.resultResponse(req.ID, toolResult(text, false)), true
}

// toolResult builds an MCP tools/call result with a single text content block.
func toolResult(text string, isError bool) map[string]any {
	return map[string]any{
		"content": []map[string]any{
			{"type": "text", "text": text},
		},
		"isError": isError,
	}
}

// toolError returns a tools/call result flagged as an error (the MCP convention
// for tool-level failures, distinct from protocol-level JSON-RPC errors).
func (s *Server) toolError(id json.RawMessage, msg string) []byte {
	return s.resultResponse(id, toolResult("Error: "+msg, true))
}

func (s *Server) resultResponse(id json.RawMessage, result any) []byte {
	resp := map[string]any{
		"jsonrpc": "2.0",
		"result":  result,
	}
	if len(id) > 0 {
		resp["id"] = id
	}
	out, err := json.Marshal(resp)
	if err != nil {
		return s.errorResponse(id, codeInternalError, "failed to marshal result")
	}
	return out
}

func (s *Server) errorResponse(id json.RawMessage, code int, message string) []byte {
	resp := map[string]any{
		"jsonrpc": "2.0",
		"error":   rpcError{Code: code, Message: message},
	}
	if len(id) > 0 {
		resp["id"] = id
	}
	out, _ := json.Marshal(resp)
	return out
}

File: experiments/adrianco/experiment-7/brazil/runs/language=clojure_model=claude-opus-4-8-fast/rep1/src/soccer/mcp.clj
;; =============================================================================
;; soccer.mcp — Model Context Protocol server (stdio / JSON-RPC 2.0)
;; -----------------------------------------------------------------------------
;; Project: brazilian-soccer-mcp
;;
;; Context:
;;   Entry point.  Speaks the MCP stdio transport: newline-delimited JSON-RPC
;;   2.0 messages on stdin/stdout (logs go to stderr so they never corrupt the
;;   protocol stream).  Implements the handshake (initialize /
;;   notifications/initialized), tool discovery (tools/list) and tool execution
;;   (tools/call) by delegating to soccer.tools.
;;
;;   The dataset is loaded once at startup (soccer.data/load-db) and kept in
;;   memory, so every tool call is a fast in-process query.
;;
;; Run:  clojure -M:run         (then talk JSON-RPC over stdin/stdout)
;; =============================================================================
(ns soccer.mcp
  (:require [clojure.data.json :as json]
            [clojure.java.io :as io]
            [clojure.string :as str]
            [soccer.data :as data]
            [soccer.tools :as tools])
  (:gen-class))

(def ^:const protocol-version "2024-11-05")

(defn- log [& xs]
  (binding [*out* *err*] (apply println xs) (flush)))

;; --- response builders ------------------------------------------------------

(defn- result-msg [id result]
  {:jsonrpc "2.0" :id id :result result})

(defn- error-msg [id code message]
  {:jsonrpc "2.0" :id id :error {:code code :message message}})

(defn- text-content [s]
  {:content [{:type "text" :text s}]})

;; --- request handling -------------------------------------------------------

(defn handle-request
  "Handle a single decoded JSON-RPC request map against `db`.
   Returns a response map, or nil for notifications (no reply expected)."
  [db {:keys [method id params]}]
  (case method
    "initialize"
    (result-msg id {:protocolVersion protocol-version
                    :capabilities {:tools {}}
                    :serverInfo {:name "brazilian-soccer-mcp"
                                 :version "1.0.0"}})

    "notifications/initialized" nil
    "initialized" nil
    "ping" (result-msg id {})

    "tools/list"
    (result-msg id {:tools (vec tools/tool-specs)})

    "tools/call"
    (let [tool-name (or (get params "name") (get params :name))
          args      (or (get params "arguments") (get params :arguments) {})]
      (try
        (let [text (tools/call-tool db tool-name args)]
          (result-msg id (text-content text)))
        (catch clojure.lang.ExceptionInfo e
          (if (= ::tools/unknown-tool (:type (ex-data e)))
            (error-msg id -32601 (.getMessage e))
            (result-msg id (assoc (text-content (str "Error: " (.getMessage e)))
                                  :isError true))))
        (catch Exception e
          (log "tool error:" (.getMessage e))
          (result-msg id (assoc (text-content (str "Error: " (.getMessage e)))
                                :isError true)))))

    ;; unknown method
    (when id (error-msg id -32601 (str "Method not found: " method)))))

;; --- stdio loop -------------------------------------------------------------

(defn serve
  "Run the JSON-RPC stdio loop reading from `in` and writing to `out`."
  [db in out]
  (let [w (io/writer out)]
    (binding [*out* w]
      (doseq [line (line-seq (io/reader in))]
        (when-not (str/blank? line)
          (let [resp (try
                       (handle-request db (json/read-str line :key-fn keyword))
                       (catch Exception e
                         (log "parse/handle error:" (.getMessage e))
                         (error-msg nil -32700 "Parse error")))]
            (when resp
              (println (json/write-str resp))
              (flush))))))))

(defn -main [& _]
  (log "brazilian-soccer-mcp: loading datasets...")
  (let [db (data/load-db)]
    (log (format "brazilian-soccer-mcp: ready (%d matches, %d players)"
                 (count (:matches db)) (count (:players db))))
    (serve db System/in System/out)))
